keps/sig-api-machinery/20190820-master-service-of-type-externalname.md

```diff
@@ -0,0 +1,302 @@
+
+The master service reconciler in kube-apiserver is extended to support services of type `ExternalName`.
+
+Example:
+
+Given a kube-apiserver advertised with host `api.127.0.0.1.nip.io` and port `443`, it will reconcile the master service with the following content:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    component: apiserver
+    provider: kubernetes
+  name: kubernetes
+  namespace: default
+spec:
+  clusterIP: ""
+  type: ExternalName
+  externalName: api.127.0.0.1.nip.io
+  ports:
+  - name: https
+    port: 443
+    protocol: TCP
+    targetPort: 443
+```
+
+It'll also delete the corresponding `Endpoints` called `kubernetes` from the `default` namespace as it's not needed for `Services` of type `ExternalName`.
+
+To provide a backwards-compatible way for cluster operators to enable/disable/configure this feature a new flag is going to be introduced:
+
+- `--kubernetes-service-external-name` with the following description:
+
+  ```text
+  The hostname on which to advertise the apiserver to members of the cluster.
+  This address must be reachable by the rest of the cluster.
+  The master service will be of type ExternalName, using this as the value of the externalName.
+  If blank, --advertise-address is going to be used and the master service will be of type ClusterIP.
+  Useful if the apiserver is running behind a load balancer which does not have a static ip.
+  ```
+
+### kubelet
+
+No changes to kubelet flags are needed.
+
+The environment variable injection mechanism is going to be changed to allow the master service to be of type `ExternalName`.
+
+Example:
+
+Given a kube-apiserver advertised with host `api.127.0.0.1.nip.io` and port `443` when executing `env` command in a Pod's container, the following should be returned:
+
+```text
+KUBERNETES_SERVICE_PORT=443
+KUBERNETES_PORT=tcp://api.127.0.0.1.nip.io:443
+KUBERNETES_PORT_443_TCP_ADDR=api.127.0.0.1.nip.io
+KUBERNETES_PORT_443_TCP_PORT=443
+KUBERNETES_PORT_443_TCP_PROTO=tcp
+KUBERNETES_SERVICE_PORT_HTTPS=443
+KUBERNETES_PORT_443_TCP=tcp://api.127.0.0.1.nip.io443
+KUBERNETES_SERVICE_HOST=api.127.0.0.1.nip.io
+
+# container-specific variables
+HOSTNAME=busybox
+SHLVL=1
+HOME=/root
+TERM=xterm
+PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+```
+
+### Implementation Notes
+
+While the `--external-hostname` flag already exists in kube-apiserver, it's documented as targeting mainly external (not in-cluster) clients:
+
+```text
+--external-hostname string
+  The hostname to use when generating externalized URLs for this master (e.g. Swagger API Docs).
+```
+
+and in some cases / network configurations, the API server might be advertised as `api.external.example.com` to (human) end-users, but in the same time it can be only reachable via `api.internal.example.com` from in-cluster clients / Pods.
+
+### Risks and Mitigations
+
+This feature requires a running DNS solution such as CoreDNS for connectivity to the API server.
```

Comment:
I see a possible chicken an egg problem, because the DNS is exposed as a service and the DNS pods need to get to the API first. It seems that the DNS pods need to resolve locally the API address, right?

Reply:
This is mentioned in the `Drawback` section. https://github.com/kubernetes/enhancements/pull/1216/files/e0f108b90433dba257fb4bc95c2e169b39a98b70#diff-1a8c28c5092f18d7eae29d98e4ae0eb6R272

> in-cluster DNS solutions such as CoreDNS must have `spec.dnsPolicy: Default` to avoid chicken and egg problem. This is already the case for [CoreDNS](https://github.com/kubernetes/kubernetes/blob/1719ce7883db57d24d70605606f189469fd50b60/cluster/addons/dns/coredns/coredns.yaml.base#L164) and [KubeDNS](https://github.com/kubernetes/kubernetes/blob/1719ce7883db57d24d70605606f189469fd50b60/cluster/addons/dns/kube-dns/kube-dns.yaml.base#L217).

Reply:
ups, sorry, totally missed that part